internal/plugin/plugin.go

```diff
@@ -93,12 +95,39 @@ func NewPluginClient(connector string, version string, logger hclog.Logger) (Plu
 		return nil, fmt.Errorf("the plugin (%s) doesn't correctly implement the necessary interfaces", connector)
 	}
 
+	currentVersion := version.CurrentCliVersionInformation{}
+	if _, err = isPluginCompatibleWithCli(&currentVersion, is, logger); err != nil {
+		logger.Error("Plugin not compatible: %s", err.Error())
+		return nil, err
+	}
+
 	info := is.PluginInfo()
 	logger.Debug("Using plugin: " + info.String())
 
 	return pci, nil
 }
 
+func isPluginCompatibleWithCli(currentCliInformation version.CliVersionInformation, pluginInfo plugin2.Info, logger hclog.Logger) (bool, error) {
+	if currentCliInformation.GetCliVersion().Equal(version.DevVersion) {
+		logger.Warn("Skip plugin compatibility check for dev version")
+		return true, nil
+	}
+
+	//Check if current CLI version is compatible with CLI version used at plugin compile time (current version > plugin compiled version)
+	pluginCompiledVersion := pluginInfo.CliBuildVersion()
+	if currentCliInformation.CliPluginConstraint().Check(&pluginCompiledVersion) {
+		return true, nil
+	}
+
+	//Check if plugin CLI version is compatible with current CLI version (current version < plugin compiled version)
+	pluginConstraint := pluginInfo.PluginCliConstraint()
+	if pluginConstraint.Check(currentCliInformation.GetCliVersion()) {
+		return true, nil
+	}
+
+	return false, fmt.Errorf("plugin (%s) doesn't support the current CLI version (%s)", pluginInfo.PluginInfo().Version.String(), version.GetCliVersion().String())
```

Comment:
We should give as much information here as possible. So it's easy for the user to solve. e.g. provide possible solutions.